src/rpc.c

```diff
@@ -161,6 +163,364 @@ gboolean ctxusb_daemon_new_vm(CtxusbDaemonObject *this,
+    {
+      g_hash_table_insert(rule_sysattrs,
+          g_strdup(*dict),
+          g_strdup(*(dict + 1)));
+      dict += 2;
+    }
+  }
+
+  dict = rule->dev_properties;
+  if (dict != NULL)
+  {
+    while (*dict != NULL && *(dict + 1) != NULL)
+    {
+      g_hash_table_insert(rule_properties,
+          g_strdup(*dict),
+          g_strdup(*(dict + 1)));
+      dict += 2;
+    }
+  }
+
+  if (rule->dev_vendorid == 0x0)
+    *OUT_vendor_id = g_strdup("");
+  else
+    *OUT_vendor_id = g_strdup_printf("%04X", rule->dev_vendorid);
+
+  if (rule->dev_deviceid == 0x0)
+    *OUT_device_id = g_strdup("");
+  else
+    *OUT_device_id = g_strdup_printf("%04X", rule->dev_deviceid);
+
+  *OUT_command = policy_parse_command_enum(rule->cmd);
+  *OUT_description = g_strdup(rule->desc);
+  *OUT_serial_number = g_strdup(rule->dev_serial);
+  *OUT_sysattrs = rule_sysattrs;
+  *OUT_udev_properties = rule_properties;
+  *OUT_vm_uuid = g_strdup(rule->vm_uuid);
+  return TRUE;
+}
+
+gboolean ctxusb_daemon_policy_list(CtxusbDaemonObject *this,
+    GArray* *OUT_rules,
+    GError** error)
+{
+  uint16_t *rule_list = NULL;
+  size_t size = 0;
+
+  GArray *index_array = g_array_new(FALSE, TRUE, sizeof(gint));
+  *OUT_rules = index_array;
+
+  policy_list_rules(&rule_list, &size);
+
+  for (size_t index=0; index < size; index++)
+  {
+    gint value = (gint)(rule_list[index]);
+    g_array_append_val(index_array, value);
+  }
+  if (size != 0 && rule_list != NULL) free(rule_list);
+
+  return TRUE;
+}
+
+gboolean ctxusb_daemon_policy_get_rules(CtxusbDaemonObject *this,
+    GPtrArray* *OUT_rule_set,
+    GError** error)
+{
+  uint16_t *rule_list = NULL;
+  size_t size = 0;
+
+  GPtrArray* response = g_ptr_array_new();
+  *OUT_rule_set = response;
+
+  policy_list_rules(&rule_list, &size);
+
+  for (size_t index=0; index < size; index++)
+  {
+    gint pos = (gint)rule_list[index];
+    char *command = NULL;
+    char *description = NULL;
+    char *vendor_id = NULL;
+    char *device_id = NULL;
+    char *serial_number = NULL;
+    GHashTable *sysattrs;
+    GHashTable *udev_properties;
+    char* vm_uuid = NULL;
+
+    if (! ctxusb_daemon_policy_get_rule(
+        this,
+        pos,
+        &command,
+        &description,
+        &vendor_id,
+        &device_id,
+        &serial_number,
+        &sysattrs,
+        &udev_properties,
+        &vm_uuid,
+        error)) continue;
+
+    GValue *value = g_new0(GValue, 1);
+    g_value_init(value, DBUS_RULE_STRUCT);
+    g_value_take_boxed(value,
+        dbus_g_type_specialized_construct(DBUS_RULE_STRUCT));
+    dbus_g_type_struct_set(value,
+        0, pos,
+        1, command,
+        2, description,
+        3, vendor_id,
+        4, device_id,
+        5, serial_number,
+        6, sysattrs,
+        7, udev_properties,
+        8, vm_uuid,
+        G_MAXUINT);
+    g_ptr_array_add(response, g_value_get_boxed(value));
+
+    g_free(command);
+    g_free(vendor_id);
```

Comment:
Missing `g_free(description)` here since it is `strdup`'ed in `ctxusb_daemon_policy_get_rule`?

Reply:
Silly mistake, thanks for catching. Pushed rebased fix for the missing field.

Reply:
Thanks!